File: 20. Valid Parentheses/20. Valid Parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function (s) {
  const parentheses = new Map([
    [")", "("],
    ["]", "["],
    ["}", "{"],
  ]);
  const stack = [];

  for (const char of s) {
    if (parentheses.has(char)) {
      if (!stack) {
        return false;
      }

      const popped = stack.pop();
      if (popped !== parentheses.get(char)) {
        return false;
      }
    } else {
      stack.push(char);
    }
  }

  return stack.length === 0;
};

File: 20. Valid Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        parentheses = {")": "(", "]": "[", "}": "{"}
        stack = []

        for char in s:
            if char not in parentheses:
                stack.append(char)
            else:
                if not stack or stack and stack[-1] != parentheses[char]:
                    return False
                stack.pop()

        return len(stack) == 0
